evm/p2p/discovery.py

```diff
@@ -0,0 +1,320 @@
+        # datagram must be 1280 bytes, so when sending neighbours packets we must include up to
+        # _max_neighbours_per_packet and if there's more than that split them across multiple
+        # packets.
+        if self._max_neighbours_per_packet_cache is not None:
+            return self._max_neighbours_per_packet_cache
+        # Use an IPv6 address here as we're interested in the size of the biggest possible node
+        # representation.
+        addr = kademlia.Address('::1', 30303, 30303)
+        node_data = addr.to_endpoint() + [b'\x00' * (kademlia.k_pubkey_size // 8)]
+        neighbours = [node_data]
+        expiration = rlp.sedes.big_endian_int.serialize(int(time.time() + EXPIRATION))
+        payload = rlp.encode([neighbours] + [expiration])
+        while HEAD_SIZE + len(payload) <= 1280:
+            neighbours.append(node_data)
+            payload = rlp.encode([neighbours] + [expiration])
+        self._max_neighbours_per_packet_cache = len(neighbours) - 1
+        return self._max_neighbours_per_packet_cache
+
+    def listen(self, loop):
+        return loop.create_datagram_endpoint(
+            lambda: self, local_addr=(self.address.ip, self.address.udp_port))
+
+    def connection_made(self, transport):
+        self.transport = transport
+
+    @asyncio.coroutine
+    def bootstrap(self):
+        while self.transport is None:
+            # FIXME: Instead of sleeping here to wait until connection_made() is called to set
+            # .transport we should instead only call it after we know it's been set.
+            yield from asyncio.sleep(1)
+        self.logger.debug("boostrapping with {}".format(self.bootstrap_nodes))
+        yield from self.kademlia.bootstrap(self.bootstrap_nodes)
+
+    def datagram_received(self, data, addr):
+        ip_address, udp_port = addr
+        self.receive(kademlia.Address(ip_address, udp_port), data)
+
+    def error_received(self, exc):
+        self.logger.error('error received: {}'.format(exc))
+
+    def send(self, node, message):
+        self.transport.sendto(message, (node.address.ip, node.address.udp_port))
+
+    def stop(self):
+        self.logger.info('stopping discovery')
+        self.transport.close()
+
+    def receive(self, address, message):
+        try:
+            remote_pubkey, cmd_id, payload, mdc = _unpack(message)
+            # Note: as of discovery version 4, expiration is the last element for all
+            # packets. This might not be the case for a later version, but just popping
+            # the last element is good enough for now.
+            expiration = rlp.sedes.big_endian_int.deserialize(payload.pop())
+            if time.time() > expiration:
+                raise PacketExpired()
+        except DefectiveMessage as e:
+            self.logger.error('error unpacking message: {}'.format(e))
+            return
+
+        cmd = CMD_ID_MAP[cmd_id]
+        if len(payload) != cmd.elem_count - 1:
```

Comment:
Why is this `cmd.elem_count - 1`?  What's the extra *thing* in the payload?

Reply:
All packages include an expiration as the last element, but a few lines above we pop() that out of the payload, that's why we compare the # of elements in the payload to elem_count-1. I'll add a comment to make that clearer, hope it's enough

Reply:
Hrm, this feels weird.  In that case, it seems like we shouldn't be popping it off of the payload.  Also, if we don't pop it from the payload we'll remove a bit of mutation.  thoughts?

Reply:
Yeah, I've changed it to not pop(), and also changed the handler methods to unpack the payload, with a comment to make it clear what they expect to be in the payload